### PR TITLE
Redirect API decoupling

### PR DESCRIPTION
**Pull request by JasonChong96**
## Problem

The redirect API and model components are highly coupled. This makes it difficult to inject mocks for automated user testing. Prerequisite for solving #70  

## Solution

Decouple redirect API from Redis and Sequelize using InversifyJS. Split up code in the model component.

**Improvements**:

- `redirect.ts` has no knowledge of cache and database implementation
- `index.ts` in `server/model` has been split into `user.ts` and `url.ts`

## Deploy Notes

**New dependencies**:

- `inversify`: Dependency injection library
- `reflect-metadata`: Allow the use of annotations in js/ts

### CODE REVIEW

File: src/server/models/url.ts [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+}, {
+  hooks: {
+    afterCreate: async (url: UrlType, options) => {
+      if (!options.transaction) {
+        return Promise.reject(new Error('URL creation must be wrapped in a transaction'))
+      }
+      // eslint-disable-next-line no-use-before-define
+      await writeToUrlHistory(
+        url,
+                options as Sequelize.CreateOptions & { transaction: Sequelize.Transaction },
```
Comment (yong-jie) [inline src/server/models/url.ts:84]: indentation looks a bit broken in many places like this. did any lint errors pop up for you? if not then something is fishy about our lint config

Comment (JasonChong96) [reply src/server/models/url.ts:84]: There aren't any errors. Seems to have happened with VSCode formatting using our lint config, will check if it's my settings. I'll manually fix these first.

File: src/server/inversify.config.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import { Container } from 'inversify'
+import { AnalyticsLogger } from '../types/server/api/analytics'
+import UrlCacheRedis from './api/cache/url'
+import UrlRepositorySequelize from './api/repositories/url'
+import GaLogger from './api/analytics/galogger'
+
+const Symbols = {
+  urlCache: Symbol.for('urlCache'),
+  urlRepository: Symbol.for('urlRepository'),
+  aLogging: Symbol.for('aLogging'),
```
[the remark this answers is not in the document]
Comment (yong-jie) [reply src/server/inversify.config.ts:10]: edit: okay this was probably analytics logging. i think this variable name might not be verbose enough by itself to convey what it is. if you opt to keep, its fine too

File: src/server/api/cache/url.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
+import { injectable } from 'inversify'
+import { redirectClient } from '../../redis'
+import { logger, redirectExpiry } from '../../config'
+import { NotFoundError } from '../../util/error'
+
+/* eslint class-methods-use-this: ["error", { "exceptMethods":
+    ["getLongUrlFromCache", "cacheShortUrl"] }] */
```
Comment (yong-jie) [inline src/server/api/cache/url.ts:7]: same thing here. indent is 4 spaces for this comment. was it supposed to be 2?

File: src/types/server/api/analytics.d.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import Express from 'express'
+
+export interface AnalyticsLogger {
+    /**
+     *
```
[the remark this answers is not in the document]
Comment (liangyuanruo) [reply src/types/server/api/analytics.d.ts:5]: Created a new issue in #74 to resolve formatting issues

File: src/server/api/repositories/url.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
+import { injectable } from 'inversify'
+import { Url } from '../../models/url'
+import { ACTIVE } from '../../models/types'
+import { NotFoundError } from '../../util/error'
+
+@injectable()
+/* eslint class-methods-use-this: ["error", { "exceptMethods":
+    ["getLongUrlFromDatabase", "incrementClick"] }] */
```
[the remark this answers is not in the document]
Comment (JasonChong96) [reply src/server/api/repositories/url.ts:8]: my bad somehow my VSCode wasn't properly synced to the project configurations.

File: src/server/inversify.config.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (JasonChong96) [reply src/server/inversify.config.ts:10]: Will change it to analyticsLogging

File: src/types/server/api/analytics.d.ts [hunk 6]
```diff
@@ -0,0 +1,17 @@
+import Express from 'express'
+
+export interface AnalyticsLogger {
+    /**
```
[the remark this answers is not in the document]
Comment (JasonChong96) [reply src/types/server/api/analytics.d.ts:4]: That makes sense, I'll move them to a regular `.ts` file

**Comment by JasonChong96** [comment]
Pushed a new commit to address changes

File: src/types/server/api/analytics.d.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (JasonChong96) [reply src/types/server/api/analytics.d.ts]: Found an issue with configurations, created a new issue in #75 to resolve

**Comment by yong-jie** [comment]
Remember to use the squash and merge option

**Pull request merged**